File: menubar-app/Sources/BurnBarCore/Forecast.swift
import Foundation

/// Swift mirror of the sidecar's tested forecast/threshold math, extended with
/// configurable thresholds (the sidecar handles Codex; this handles Claude).
public struct PercentSample: Equatable, Sendable {
    public let at: Date
    public let percent: Double
    public init(at: Date, percent: Double) {
        self.at = at
        self.percent = percent
    }
}

public enum Forecast {
    /// Linear projection of when the window hits 100%. nil when flat/declining
    /// or fewer than 2 samples inside the lookback.
    public static func limitHit(
        samples: [PercentSample],
        now: Date,
        lookback: TimeInterval = 3600
    ) -> Date? {
        let recent = samples
            .filter { now.timeIntervalSince($0.at) <= lookback && $0.percent.isFinite }
            .sorted { $0.at < $1.at }
        guard recent.count >= 2, let first = recent.first, let last = recent.last else { return nil }
        if last.percent >= 100 { return now }
        let dt = last.at.timeIntervalSince(first.at)
        let dp = last.percent - first.percent
        guard dt > 0, dp > 0 else { return nil }
        let rate = dp / dt
        return last.at.addingTimeInterval((100 - last.percent) / rate)
    }
}

/// Rising-edge detection with CONFIGURABLE thresholds and reset awareness.
public struct ConfigurableAlertEdge: Sendable {
    public var thresholds: [Double]
    private var previous: Double?

    public init(thresholds: [Double] = [80, 95]) {
        self.thresholds = thresholds.sorted()
    }

    public mutating func feed(_ percent: Double?) -> [(kind: String, level: Int?, percent: Double)] {
        guard let percent, percent.isFinite else { return [] }
        defer { previous = percent }
        guard let prev = previous else { return [] }
        let lowest = thresholds.first ?? 80
        if prev - percent >= 30 || (prev >= lowest && percent < lowest) {
            return [("reset", nil, percent)]
        }
        return thresholds.compactMap { level in
            prev < level && percent >= level ? ("threshold", Int(level), percent) : nil
        }
    }
}

/// One-shot "shoulder tap": fires when a limit is projected to hit the cap
/// within `leadTime`, at most once per window climb (re-arms on reset/decline).
public struct ForecastAlarm: Sendable {
    public var leadTime: TimeInterval
    private var armed = true

    public init(leadTime: TimeInterval = 30 * 60) {
        self.leadTime = leadTime
    }

    public mutating func feed(forecastHit: Date?, percent: Double?, now: Date) -> Bool {
        if let percent, percent < 50 { armed = true } // window rolled over / calmed down
        guard armed, let hit = forecastHit else { return false }
        let dt = hit.timeIntervalSince(now)
        guard dt > 0, dt <= leadTime else { return false }
        armed = false
        return true
    }
}

File: menubar-app/sidecar/src/forecast.ts
/**
 * Pure math for "at this pace you hit the cap at 6:40 PM" and for alerting.
 * Kept runtime-free (no Date.now, no I/O) so it unit-tests exactly.
 */

export interface PercentSample {
  /** ms epoch */
  at: number;
  percent: number;
}

/**
 * Linear pace forecast: from the oldest→newest samples inside `lookbackMs`,
 * project when percent reaches 100. Returns the projected ms epoch, or null
 * when there's no meaningful upward burn (fewer than 2 samples, flat, or
 * declining — a decline means the window rolled over).
 */
export function forecastLimitHit(
  samples: PercentSample[],
  nowMs: number,
  lookbackMs = 60 * 60 * 1000,
): number | null {
  const recent = samples
    .filter((s) => nowMs - s.at <= lookbackMs && Number.isFinite(s.percent))
    .sort((a, b) => a.at - b.at);
  if (recent.length < 2) return null;
  const first = recent[0];
  const last = recent[recent.length - 1];
  if (last.percent >= 100) return nowMs;
  const dt = last.at - first.at;
  const dp = last.percent - first.percent;
  if (dt <= 0 || dp <= 0) return null;
  const ratePerMs = dp / dt;
  if (ratePerMs <= 0) return null;
  return Math.round(last.at + (100 - last.percent) / ratePerMs);
}

export type AlertKind = "threshold" | "reset";

export interface LimitAlert {
  kind: AlertKind;
  /** The threshold crossed (80 or 95) for kind=threshold. */
  level?: number;
  percent: number;
}

export const ALERT_THRESHOLDS = [80, 95] as const;

/**
 * Rising-edge threshold detection with reset awareness. Feed consecutive
 * percent readings; emits each threshold at most once per climb, and a single
 * "reset" when the window rolls over (a drop of ≥30 points, or crossing back
 * below the lowest threshold after having been above it).
 */
export function detectAlerts(prevPercent: number | null, percent: number): LimitAlert[] {
  if (!Number.isFinite(percent)) return [];
  const alerts: LimitAlert[] = [];
  if (prevPercent !== null && Number.isFinite(prevPercent)) {
    const dropped = prevPercent - percent >= 30;
    const fellBelowAll = prevPercent >= ALERT_THRESHOLDS[0] && percent < ALERT_THRESHOLDS[0];
    if (dropped || fellBelowAll) {
      alerts.push({ kind: "reset", percent });
      return alerts;
    }
  }
  for (const level of ALERT_THRESHOLDS) {
    const wasBelow = prevPercent === null ? true : prevPercent < level;
    if (wasBelow && percent >= level) alerts.push({ kind: "threshold", level, percent });
  }
  // Never fire a threshold on the very first sample — there's no edge yet.
  if (prevPercent === null) return [];
  return alerts;
}
